src/openassetio-python/bridge/src/python/converter.cpp

```diff
@@ -0,0 +1,105 @@
+#include <openassetio/hostApi/HostInterface.hpp>
+#include <openassetio/hostApi/Manager.hpp>
+#include <openassetio/hostApi/ManagerFactory.hpp>
+#include <openassetio/hostApi/ManagerImplementationFactoryInterface.hpp>
+#include <openassetio/log/ConsoleLogger.hpp>
+#include <openassetio/log/LoggerInterface.hpp>
+#include <openassetio/log/SeverityFilter.hpp>
+#include <openassetio/managerApi/Host.hpp>
+#include <openassetio/managerApi/HostSession.hpp>
+#include <openassetio/managerApi/ManagerInterface.hpp>
+#include <openassetio/managerApi/ManagerStateBase.hpp>
+
+// Private headers
+#include <openassetio/private/python/pointers.hpp>
+
+namespace py = pybind11;
+
+namespace openassetio {
+inline namespace OPENASSETIO_CORE_ABI_VERSION {
+namespace python::converter {
+
+template <typename T>
+PyObject* castToPyObject(const T& objectPtr) {
+  // Stash the errors for restoring after cast.
+  // An error_scope calls PyErr_Fetch (which clears the active error),
+  // and stashes them, restoring the error state when it falls out of
+  // scope
+  const py::error_scope previousErrorState;
+
+  if (objectPtr == nullptr) {
+    throw std::invalid_argument("objectPtr cannot be null");
+  }
+
+  py::object pybindObj = py::cast(objectPtr);
+
+  // Check to see if the py::cast has spawned a python error
+  // We use another error_scope here for convenience of access, relying
+  // on the fact that the original error scope will be destructed last,
+  // resetting to the initial error.
+  const py::error_scope castErrorState;
+
+  if (castErrorState.value != nullptr) {
+    throw std::runtime_error(py::str(castErrorState.value));
+  }
+
+  // `release()` to avoid decrementing the PyObject refcount on leaving
+  // this function. We explicitly want +1 refcount.
+  return pybindObj.release().ptr();
+}
+
+template <typename T>
+typename T::Ptr castFromPyObject(PyObject* pyObject) {
+  if (pyObject == nullptr) {
+    throw std::invalid_argument("Attempting to cast a nullptr PyObject");
+  }
+
+  // get Python object.
+  const auto pyInstance = py::reinterpret_borrow<py::object>(pyObject);
+
+  try {
+    // Extract the underlying C++ base class pointer.
+    auto* cppInstancePtr = py::cast<T*>(pyInstance);
+
+    // Use aliasing constructor, linking Python instance and C++ instance
+    // lifetimes via PyObject refcount.
+    return pointers::createPyRetainingPtr<typename T::Ptr>(pyInstance, cppInstancePtr);
+  } catch (const py::cast_error& castError) {
+    // Rethrow here to avoid bleeding pybind exception dependencies.
+    throw std::invalid_argument(castError.what());
+  }
+}
+
+// To/from explicit specialization convenience.
+// As these specializations are being used across library boundaries,
+// we must export them to prevent undefined-symbol link errors.
+#define OPENASSETIO_SPECIALIZE_PYTHON_CONVERSIONS(Class)                          \
+  template OPENASSETIO_PYTHON_BRIDGE_EXPORT PyObject* castToPyObject<Class::Ptr>( \
```

Comment:
[question] Do we need to consider `ConstPtr` here at all? I know pybind erases constness, I guess C++ is happy to pop the non-const one into a const one on assignment?

Reply:
I think pybind11 casting away constness is a bit eyebrow-raising.  I think allowing a cast from a `ConstPtr` to a non-const `PyObject*`, effectively providing a backdoor way to bypass constness, possibly without realising it, would just be confusing.

Reply:
Cool. I take it its a compile error if you you try to cast a ConstPtr to py? Be good to note this situation in the docs - seems useful.

Reply:
72bf6bbc0faa10c76baf4052c50dd44f4acff929